### ISSUE TITLE
"autosave: true" on belongs_to of join model causes invalid records to be saved

### ISSUE THREAD

**Issue by lcreid** [rails/rails#28080]
We have an application that has a `has_many through:` association. We need to be able to create instances of the models and their associations in memory, work with them, and only save or update them in certain situations. 

We found that if we have `autosave: true` on the `belongs_to` association in the join model, saving one of the other models causes Rails to save the join model without some of the required IDs. See the test code and description below for more details.

I will look into the Rails code today to see if I can come up with a fix. If this is expected behaviour, please let me know. In that case, I would be happy to submit some enhancements to the documentation at least.

### Steps to reproduce

```
begin
  require "bundler/inline"
rescue LoadError => e
  $stderr.puts "Bundler version 1.10 or later is required. Please update your Bundler"
  raise e
end

gemfile(true) do
  source "https://rubygems.org"
  gem "rails", github: "rails/rails"
  gem "arel", github: "rails/arel"
  gem "sqlite3"
end

require "active_record"
require "minitest/autorun"
require "logger"

# This connection will do for database-independent bug reports.
ActiveRecord::Base.establish_connection(adapter: "sqlite3", database: ":memory:")
ActiveRecord::Base.logger = Logger.new(STDOUT)

ActiveRecord::Schema.define do
  create_table :courses do |t|
    t.string :name
  end

  create_table :sessions do |t|
    t.date :start_date
    t.date :end_date
    t.string :name
  end

  create_table :sections do |t|
    t.string :short_name
    t.belongs_to :session, foreign_key: true
    t.belongs_to :course, foreign_key: true
  end
end

##
# Represents a course.
class Course < ActiveRecord::Base
  has_many :sections, inverse_of: :course, autosave: true, dependent: :destroy
  has_many :sessions, through: :sections
end

##
# Represents a period during which the instituation offers courses.
class Session < ActiveRecord::Base
  has_many :sections, inverse_of: :session, autosave: true, dependent: :destroy
  has_many :courses, through: :sections
end

##
# Represents a course for a particular group of students,
# given at a particular session. There may be mulitple instances of a course
# given during a session.
class Section < ActiveRecord::Base
  # Remove ", autosave: truee" from the following two lines to make
  # all the tests pass.
  belongs_to :session, inverse_of: :sections, autosave: true
  belongs_to :course, inverse_of: :sections, autosave: true
end

class BugTest < Minitest::Test
  def test_belongs_to_autosave_true_works_for_one_association
    # config.logger = Logger.new(STDOUT)
    @cs180 = Course.new(name: "CS180")
    @fall = Session.new(name: "Fall",
                        start_date: "2017-09-01",
                        end_date: "2017-12-31")
    @fall.sections << (@cs180a = @cs180.sections.build(short_name: "A"))

    assert @cs180.save

    @cs180.reload
    @fall.reload
    assert_equal [@cs180a].sort_by(&:id), @cs180.sections.sort_by(&:id)
    assert_equal [@cs180a].sort_by(&:id), @fall.sections.sort_by(&:id)
  end

  def test_belongs_to_autosave_true_fails_for_two_associations
    @cs180 = Course.new(name: "CS180")
    @fall = Session.new(name: "Fall",
                        start_date: "2017-09-01",
                        end_date: "2017-12-31")
    @fall.sections << (@cs180a = @cs180.sections.build(short_name: "A"))
    @fall.sections << (@cs180b = @cs180.sections.build(short_name: "B"))

    assert @cs180.save

    puts "\n\n"
    puts "NOTE nil IDs shouldn't have passed validation"
    Section.all.each { |s| puts "SECTIONS AFTER SAVE: #{s.inspect}" }
    puts "\n\n"

    @cs180.reload
    @fall.reload
    assert_equal [@cs180a, @cs180b].sort_by(&:id),
                 @cs180.sections.sort_by(&:id)
    assert_equal [@cs180a, @cs180b].sort_by(&:id),
                 @fall.sections.sort_by(&:id)
  end

  def test_belongs_to_autosave_true_fails_for_three_associations
    @cs180 = Course.new(name: "CS180")
    @fall = Session.new(name: "Fall",
                        start_date: "2017-09-01",
                        end_date: "2017-12-31")
    @fall.sections << (@cs180a = @cs180.sections.build(short_name: "A"))
    @fall.sections << (@cs180b = @cs180.sections.build(short_name: "B"))
    @fall.sections << (@cs180c = @cs180.sections.build(short_name: "C"))

    assert @cs180.save

    puts "\n\n"
    puts "NOTE nil IDs shouldn't have passed validation"
    Section.all.each { |s| puts "SECTIONS AFTER SAVE: #{s.inspect}" }
    puts "\n\n"

    @cs180.reload
    @fall.reload
    assert_equal [@cs180a, @cs180b, @cs180c].sort_by(&:id),
                 @cs180.sections.sort_by(&:id)
    assert_equal [@cs180a, @cs180b, @cs180c].sort_by(&:id),
                 @fall.sections.sort_by(&:id)
  end
end
```

### Expected behavior
Saving the Course object should save the Session model, and the join model with IDs for both the Courses and the Sessions.

### Actual behavior
The first join model is saved with the correct IDs, but subsequent join models are saved with the `session_id` as nil. A nil `session_id` should fail validation in Rails 5, but no exception is thrown, either.

### System configuration
**Rails version**: 5.0.1, 5.1.0.alpha

**Ruby version**: 2.3.1p112, 2.4.0p0

**Comment by al2o3cr** [comment]
I believe I know the issue - the `autosave` callbacks of the various pieces are getting interleaved in a bad order. Will try to get a detailed writeup together soon.

I'm curious what the `autosave: true` on the `belongs_to` here is for; as written here, it wouldn't seem like you'd write code like:
```ruby
section = Section.find(params[:id])
section.short_name = params[:short_name]
section.course.name = params[:this_seems_odd]
section.save
```
I assume this is not the original domain model, but autosaving a `belongs_to` seems odd.

The one case I've seen a similar situation happen looked like this:
```
class Person < ActiveRecord::Base
  belongs_to :current_mailing_address, class_name: Address'
  accepts_nested_attributes_for :current_mailing_address # turns on autosave: true
 
  belongs_to :current_residence_address, class_name: 'Address'
  accepts_nested_attributes_for :current_residence_address

  has_many :addresses
  accepts_nested_attributes_for :addresses
end

class Address < ActiveRecord::Base
  belongs_to :person
end
```
(we've got a little problem with overuse of `accepts_nested_attributes_for` at work :) )
Under certain circumstances, this can result in addresses saving before the person does. That may be the same bug here, as both involve `autosave: true`.

**Comment by lcreid** [comment]
Yes, it's absolutely about the order the callbacks are called in. The callbacks start firing from the Seminar instance and go along nicely to save the first Section instance, but then they "bounce off" the Session instance and start saving the other Section instances, and that's when the Section instances get saved without the `seminar_id`.

I pushed a "fix" to a branch in my fork at https://github.com/lcreid/rails/tree/belongs_to. The "fix" is in `activerecord/lib/active_record/autosave_association.rb`, in [`save_collection_association` about line 419](https://github.com/lcreid/rails/blob/belongs_to/activerecord/lib/active_record/autosave_association.rb#L419). All tests pass with this "fix", but I don't really understand all the ramifications of it, so I wouldn't seriously propose it as a fix. At least not yet.

(The branch includes two test cases to trigger my issue.)

As to why I want to do this: Part of it comes from me telling some new Rails programmers to just trust ActiveRecord to autosave everything, and them we immediately ran into cases where it looked like I was a liar. :-) 

We also have some use cases in our current application where we need to create a number of associated models in memory, and only save in certain circumstances. We hoped that the ActiveRecord magic would ensure that everything that had to get saved, would get save. It most does, except for this case.

In short, yes, I realize we're on the hairy edge of what we probably *should* try to do with ActiveRecord, but it would be nice if this behaved in an expected way.

I will look into this further today (Pacific Time zone). Thanks for the response and encouragement.

**Comment by al2o3cr** [comment]
@lcreid I've been puttering around for a long time trying to solve the problem of getting the association save order to be more deterministic; it's a remarkable construction that (when working *correctly*) effectively does a topological sort of the object graph using `before_save` and `after_save` callbacks.

When it malfunctions, the result is totally mysterious weirdness like this, or the issues we've seen over the years with `accepts_nested_attributes_for`.

BTW, if you're building models in memory and saving them you shouldn't need `autosave: true` in 99% of cases - the behavior is implemented more as "save associated records if (we are new and autosave isn't false) OR autosave is true".

**Comment by lcreid** [comment]
Maybe the problem is that the callbacks for the collection associations aren't re-entrant, because the first callback is: https://github.com/lcreid/rails/blob/belongs_to/activerecord/lib/active_record/autosave_association.rb#L369. When the chain of `autosave: true` associations is circular, this callback gets called twise, the second time settin `@new_record_before_save` to false before the association has been fully updated with the missing association ID. This sends the code in `save_collection_association` down a path that doesn't update the missing association ID.

**Comment by lcreid** [comment]
I have a fix that makes all the tests work. I will clean up my code and push it to my fork, hopefully this evening.

**Comment by lcreid** [comment]
I have a change that seems to work. I pushed it to my fork for now: https://github.com/lcreid/rails/tree/belongs_to. @al2o3cr I'd appreciate any thoughts you have on it. I believe there are situations where the after save callback won't get called, but the before save callback was called. If so, this solution will need to be modified.

**Comment by lcreid** [comment]
No sooner did I hit send on the previous comment, and I realized a way to eliminate the explicit stack, and therefore to eliminate the potential problem of the before callback being run, but the after callback not being run. Sorry for the extra e-mails, @al2o3cr . 

I've pushed latest attempt to: https://github.com/lcreid/rails/tree/belongs_to. Again, I'd much appreciate any comments.
